### ISSUE TITLE
Teuchos:  Throwing exception when built with CMAKE_BUILD_TYPE=DEBUG and GNU 10.3 compilers

### ISSUE THREAD

**Issue by glhenni** [trilinos/Trilinos#13501]
## Bug Report
@bartlettroscoe 

### Description
Looks like there's a bug in Trilinos/develop (SHA1: 26c1ba928a87730bb2b6670623ecc1ba80b65bb2) causing an exception to be thrown. Stack trace is:
```
#0  0x00007fffc9757a7e in __cxxabiv1::__cxa_throw (obj=0x6d72b0, 
    tinfo=0x7fffc9871188 <typeinfo for std::logic_error>, 
    dest=0x7fffc976bd80 <std::logic_error::~logic_error()>)
    at /projects/aue/cee/builds/x86_64/rhel8/818788d6/gcc-10.3.0/spack/var/spack/stage/smbaxle/spack-stage-gcc-10.3.0-6ljscisq6qcb2eygwkrajkgf2sw64oqd/spack-src/libstdc++-v3/libsupc++/eh_throw.cc:78
#1  0x00007fffc9f6729c in Teuchos::RCPNodeTracer::removeRCPNode (rcp_node=0x6d4b90)
    at /gpfs/charonops/jenkins/cee/workspace/TPL-Builder-OnDemand-aue-build/Trilinos/packages/teuchos/core/src/Teuchos_RCPNode.cpp:516
#2  0x00007fffc9f679c9 in Teuchos::RCPNodeHandle::unbindOneStrong (this=0x7fffffff5628)
    at /gpfs/charonops/jenkins/cee/workspace/TPL-Builder-OnDemand-aue-build/Trilinos/packages/teuchos/core/src/Teuchos_RCPNode.cpp:695
#3  0x00007ffff7fdcd9a in Teuchos::RCPNodeHandle::unbind (this=0x7fffffff5628)
    at /projects/charon/install/TPLs/trilinos/develop/cee/rhel8/aue/openmpi-4.1.6/gnu-10.3.0/dbg/include/Teuchos_RCPNode.hpp:1030
#4  0x00007ffff7fdccd4 in Teuchos::RCPNodeHandle::~RCPNodeHandle (this=0x7fffffff5628, 
    __in_chrg=<optimized out>)
    at /projects/charon/install/TPLs/trilinos/develop/cee/rhel8/aue/openmpi-4.1.6/gnu-10.3.0/dbg/include/Teuchos_RCPNode.hpp:842
#5  0x00007ffff5b3f19c in Teuchos::RCP<Teuchos::basic_FancyOStream<char, std::char_traits<char> > >::~RCP (
    this=0x7fffffff5620, __in_chrg=<optimized out>)
    at /projects/charon/install/TPLs/trilinos/develop/cee/rhel8/aue/openmpi-4.1.6/gnu-10.3.0/dbg/include/Teuchos_RCP.hpp:282
#6  0x00007ffff5ba0e5e in Teuchos::basic_OSTab<char, std::char_traits<char> >::~basic_OSTab (
    this=0x7fffffff5620, __in_chrg=<optimized out>)
    at /projects/charon/install/TPLs/trilinos/develop/cee/rhel8/aue/openmpi-4.1.6/gnu-10.3.0/dbg/include/Teuchos_FancyOStream.hpp:692
#7  0x00007fffc9f8f1b5 in Teuchos::UnitTestRepository::runUnitTests (out=...)
    at /gpfs/charonops/jenkins/cee/workspace/TPL-Builder-OnDemand-aue-build/Trilinos/packages/teuchos/core/src/Teuchos_UnitTestRepository.cpp:279
#8  0x00007fffc9f8feb3 in Teuchos::UnitTestRepository::runUnitTestsFromMain (argc=1, argv=0x7fffffff6178)
    at /gpfs/charonops/jenkins/cee/workspace/TPL-Builder-OnDemand-aue-build/Trilinos/packages/teuchos/core/src/Teuchos_UnitTestRepository.cpp:390
#9  0x00007ffff7fd918c in main (argc=1, argv=0x7fffffff6178)
    at /home/glhenni/Projects/Charon2/tcad-charon/test/unitTests/Charon_UnitTestMain.cpp:19
```

**Comment by glhenni** [comment]
And the error I'm seeing:
```
terminate called after throwing an instance of 'std::logic_error'
  what():  /gpfs/charonops/jenkins/cee/workspace/TPL-Builder-OnDemand-aue-build/Trilinos/packages/teuchos/core/src/Teuchos_RCPNode.cpp:516:

Throw number = 1

Throw test that evaluated to true: (!rcp_node_exists)

RCPNodeTracer::removeRCPNode(node_ptr): Error, the RCPNode {address=0x6d4b90, base_obj_map_key_void_ptr=0x454545, base_obj_type_name=Teuchos::basic_FancyOStream<char, std::char_traits<char> >, map_key_void_ptr=0x454545, has_ownership=0, insertionNumber=6652848} is not found in the list of active RCP nodes being traced even though all nodes should be traced.  This should not be possible and can only be an internal programming error!
```

**Comment by bartlettroscoe** [comment]
@glhenni what previous version of Trilinos did not show this problem?  (That will bound the commits we need to look into.)

**Comment by glhenni** [comment]
Trying to determine that as I type...

**Comment by bartlettroscoe** [comment]
@glhenni, the outcome of the git bisect study will show the most recent version of Trilinos that worked, and it will also give you a recent version to revert back to while the problem is fixed offline.

Are you only doing git bisect on merge commits to the 'develop' branch?  That would be faster and has other advantages.  That will tell you which GitHub PR that broke this.

From Google search AI results for "do git bisect only on merge commits" ...

**Method 2: Using git log and grep**

Get a list of merge commits.

```
   git log --merges --pretty=format:"%H" <recent_commit> --not <last_known_good_commit> > merge_commits.txt
```

Start the bisect with the list.

```
   git bisect start <recent_commit> <last_known_good_commit> -- < merge_commits.txt
```

**Comment by glhenni** [comment]
Definitely going to keep the tip about bisecting merge commits in my back pocket. Not the first time I've had to do this.

BUT, I think I found the problem. I was linking against a DEBUG build of trilinos but my application was also linking to a RELEASE build of another library that in turn was linking against a RELEASE build of trilinos. So there was a mismatch in which build of trilinos was being used. Once I pointed my application to a DEBUG build of the other TPL everything seems to be running smoothly. 

Thanks @bartlettroscoe! Closing this...
